### PR TITLE
Add song name search functionality to lookup

### PR DESCRIPTION
**Pull request by Brainicism**
If user provides a valid URL, we assume that they are providing a youtube link and look up by youtube video ID
Otherwise, we assume they're providing a song name, and we search using song name against `available_songs.` If there are multiple results, we list the matched songs, including pagination if needed. Matches are sorted by song name length, then by views.

image

If there is only one result, we return a response similar to searching by video ID

### CODE REVIEW

File: src/commands/game_commands/lookup.ts [hunk 1]
```diff
@@ -78,263 +362,51 @@ export default class LookupCommand implements BaseCommand {
...
-
-            logger.info(
-                `${getDebugLogHeader(
-                    message
-                )} | Non-KMQ song lookup. videoID = ${videoID}.`
-            );
-        }
+        if (isValidURL(arg)) {
+            let videoID: string = null;
+            if (arg.startsWith("<") && arg.endsWith(">")) {
```
Comment (taahamahdi) [inline src/commands/game_commands/lookup.ts:371]: Is it still considered a valid url if it starts with < and ends with > ? Might need to move this before the if statement

File: src/commands/game_commands/lookup.ts [hunk 2]
```diff
@@ -39,11 +44,290 @@ const getDaisukiLink = (id: string, isMV: boolean): string => {
     return `https://kpop.daisuki.com.br/audio_videos.html?playid=${id}`;
 };
 
+async function lookupByYoutubeID(
+    message: GuildTextableMessage,
+    videoID: string,
+    guildID: string,
```
Comment (taahamahdi) [inline src/commands/game_commands/lookup.ts:50]: I think the message argument might contain the guild ID

File: src/commands/game_commands/lookup.ts [hunk 3]
```diff
@@ -39,11 +44,290 @@ const getDaisukiLink = (id: string, isMV: boolean): string => {
...
+): Promise<void> {
+    const messageContext = MessageContext.fromMessage(message);
+    const kmqSongEntries: QueriedSong[] = await dbContext
+        .kmq("available_songs")
+        .select(SongSelector.getQueriedSongFields())
+        .whereRaw(`song_name_en LIKE "%${songName}%"`)
+        .orWhereRaw(`song_name_ko LIKE "%${songName}%"`)
+        .orderByRaw("CHAR_LENGTH(song_name_en) ASC")
+        .orderBy("views", "DESC")
+        .limit(100);
```
Comment (taahamahdi) [inline src/commands/game_commands/lookup.ts:280]: Will the number of results impact performance?

File: src/commands/game_commands/lookup.ts [hunk 4]
```diff
@@ -39,11 +44,290 @@ const getDaisukiLink = (id: string, isMV: boolean): string => {
...
+
+    if (kmqSongEntries.length === 0) {
+        await sendInfoMessage(messageContext, {
+            title: "Lookup Results",
```
Comment (taahamahdi) [inline src/commands/game_commands/lookup.ts:284]: Could be localized

File: src/commands/game_commands/lookup.ts [hunk 5]
```diff
@@ -39,11 +44,290 @@ const getDaisukiLink = (id: string, isMV: boolean): string => {
...
+                value: LocalizationManager.localizer.translate(
+                    guildID,
+                    includedInOptions ? "misc.yes" : "misc.no"
+                ),
+            }
+        );
+    }
+
+    sendInfoMessage(messageContext, {
+        title: `${songName} - ${artistName}`,
```
Comment (taahamahdi) [inline src/commands/game_commands/lookup.ts:254]: Nit: The song name is not bold here but is bold in the results, which works better?

**Review by taahamahdi — CHANGES_REQUESTED** [review]

File: src/commands/game_commands/lookup.ts (hunk 3 above)
Comment (Brainicism) [reply src/commands/game_commands/lookup.ts:280]: At 100, no

**Comment by edwardhdlu** [comment]
pog

**Review by taahamahdi — APPROVED** [review]

**Comment by sabihismail** [comment]
pog

**Pull request merged**